### ISSUE TITLE
3D interpolation filters for VolumeVisual

### ISSUE THREAD

**Issue by brisvag** [vispy/vispy#2263]
I was wondering whether it was possible to implement new interpolation methods for the `VolumeVisual`, and I ended up on some of @rossant 's code from 2015 in [build_spatial_filters.py](https://github.com/vispy/vispy/blob/main/vispy/glsl/build_spatial_filters.py). It generates automatically the glsl code for 2D spatial interpolation filters (used by `Imagevisual`, for example).

I understand very little of it, but I *assume* some of it might be generalizable to 3D. Is it possible? If yes, how hard is it and where should I look for some background on it?

The original code seems to be written by @rossant , with some small changes by @kmuehlbauer . 

cc @almarklein, @campagnola, @djhoese

**Comment by kmuehlbauer** [comment]
@brisvag Yeah, small changes. I remember having a hard time to figure out those small changes :grinning:. https://github.com/vispy/vispy/pull/1069

It should be possible to generalize this to 3D. How does the Volume Visual hold the data? Kind of 3D texture?

**Comment by brisvag** [comment]
> I remember having a hard time to figure out those small changes 😃

Hehe, I noticed one of them was deep in the weeds :P Yes, it's a 3D texture. My understanding so far is that in [`filter_code()`](https://github.com/vispy/vispy/blob/185285860c0bfc55f4d44122a7f3b82b244d7d3b/vispy/glsl/build_spatial_filters.py#L114), it first builds code for 1D interpolation, and then combines it along two perpendicular dimensions to get 2D interpolation. I *guess* you could add one more section with does sort of the same but also along `z`?

**Comment by brisvag** [comment]
My thoughts go to something like this...

```glsl
vec4 filter3D_radius1(sampler3D texture, sampler3D kernel, float index, vec3 uv, vec3 pixel) {
    vec3 texel = uv/pixel - vec3(0.5, 0.5, 0.5) ;
    vec3 f = fract(texel);
    texel = (texel-fract(texel) + vec3(0.001, 0.001, 0.001)) * pixel;
    // interpolate along one dimension
    vec4 t0 = filter1D_radius1(kernel, index, f.x,
        texture3D( texture, texel + vec3(0, 0, 0) * pixel),
        texture3D( texture, texel + vec3(1, 0, 0) * pixel));
    vec4 t1 = filter1D_radius1(kernel, index, f.x,
        texture3D( texture, texel + vec3(0, 1, 0) * pixel),
        texture3D( texture, texel + vec3(1, 1, 0) * pixel));
    vec4 t2 = filter1D_radius1(kernel, index, f.x,
        texture3D( texture, texel + vec3(0, 0, 1) * pixel),
        texture3D( texture, texel + vec3(1, 0, 1) * pixel));
    vec4 t3 = filter1D_radius1(kernel, index, f.x,
        texture3D( texture, texel + vec3(0, 1, 1) * pixel),
        texture3D( texture, texel + vec3(1, 1, 1) * pixel));
    // now what? just this?
    return filter1D_radius1(kernel, index, f.y, t0, t1, t2, t3);
}
```
